java/arcs/core/analysis/DependencyNode.kt

```diff
@@ -23,117 +23,210 @@ private typealias Path = List<Identifier>
- *   ```
- *   claim output.ratio.trained derives from input.cats
- *   claim output.ratio.total derives from input.cats and derives from input.dogs
- *   claim output.family derives from input.household
- *   ```
+ *  This internal representation, in turn, can be translated into the following claims:
+ *  ```
+ *  claim output.a is equal to input.x and is influenced by input.x
+ *  claim output.b is derived from input.y and is derived from input.z and is influenced by input.x
+ *  ```
  */
 sealed class DependencyNode {
 
-  /** An unmodified input (from a handle connection) used in a Paxel [Expression]. */
-  data class Input(val path: Path = emptyList()) : DependencyNode() {
-    constructor(vararg fields: Identifier) : this(listOf(*fields))
-  }
+  /** Expresses influence on to the current [DependencyNode]. */
+  abstract fun influencedBy(influence: Set<DependencyNode>): DependencyNode
+
+  /** Abstract notion of an access path and dependency relationships. */
+  abstract class Terminal : DependencyNode() {
+    /** Path of [Identifier]s representing access to a handle connection. */
+    abstract val accessPath: Path
+
+    /** Set of [Terminal] nodes that the current node depends on. */
+    abstract val dependency: Set<Terminal>
+
+    /** Set of [Terminal] nodes that bear influence on the current node. */
+    abstract val influence: Set<Terminal>
 
-  /** Represents derivation from a group of [Input]s in an [Expression]. */
-  data class DerivedFrom(val inputs: Set<Input> = emptySet()) : DependencyNode() {
-
-    constructor(vararg paths: Path) : this(paths.map { Input(it) }.toSet())
-
-    /** Produce a new [DerivedFrom] with a flattened set of [Input]s. */
-    constructor(vararg nodes: DependencyNode) : this(flatten(*nodes))
-
-    companion object {
-      /** Flatten nested sets of [DependencyNode]s.*/
-      private fun flatten(vararg nodes: DependencyNode): Set<Input> {
-        return nodes.flatMap { node ->
-          when (node) {
-            is Input -> setOf(node)
-            is DerivedFrom -> node.inputs
-            else -> throw IllegalArgumentException(
-              "Nodes must be a 'Input' or 'DerivedFrom'."
-            )
-          }
-        }.toSet()
-      }
+    /** Gets the [Identifier] of the current node, if present (`null` otherwise). */
+    val id: Identifier?
+      get() = accessPath.last()
+
+    /** Returns unwrapped dependency node(s), or `default` if none exist. */
+    fun dependencyOrDefault(default: DependencyNode): DependencyNode = when (dependency.size) {
+      0 -> default
+      1 -> dependency.first()
+      else -> Nodes(dependency.toList())
     }
   }
 
-  /** Associates [Identifier]s with [DependencyNode]s. */
-  data class AssociationNode(
-    val associations: Map<Identifier, DependencyNode> = emptyMap()
+  /** A [Terminal] with input that has been unmodified by the expression. */
+  data class Equal(
+    override val accessPath: Path,
+    override val dependency: Set<Terminal>,
```

Comment:
`dependency` and `influence` could be members of the `DependencyNode` base class itself.

Also, the `dependency` and `influence` can be `Set<DependencyNode>` and not necessarily need to be `Terminal`?